code-studio/scripts/build-commands.js

```diff
@@ -172,39 +182,53 @@ exports.ensureDirectoryExists = function (dir) {
 };
 
 /**
- * Execute a sequence of shell commands as a build script, with useful output
- * wherever possible.
- * Will call process.exit() if the command fails for any reason.
+ * Execute a list of shell commands in sequence.
  * @param {string[]} commands - array of shell commands to be executed in sequence.
+ * @return {Promise} which resolves when all commands complete (to the stdout
+ *         value of the last command in the sequence), or rejects if any of
+ *         them fail.
  */
-exports.execute = function (commands) {
-  commands.forEach(function (command) {
-    try {
-      console.log(command);
-
-      // For documentation on synchronous execution of shell commands from node scripts, see:
-      // https://nodejs.org/docs/latest/api/child_process.html#child_process_synchronous_process_creation
-      var result = child_process.execSync(command, {
-        env: _.extend({}, process.env, {
-          PATH: './node_modules/.bin:' + process.env.PATH
-        }),
-        stdio: 'inherit'
+exports.executeSequence = function (commands) {
+  var options = getChildProcessOptions();
+  return new Promise(function (resolve, reject) {
+    var runNextCommand = function (commands) {
+      var nextCommand = commands[0];
+      console.log(nextCommand);
+      child_process.exec(nextCommand, options, function (error, stdout) {
+        if (error) {
+          reject(transformExecError(error));
+        } else if (1 === commands.length) {
+          resolve(stdout);
+        } else {
+          runNextCommand(commands.slice(1));
+        }
       });
-    } catch (e) {
-      exports.exitWithError(e);
-    }
+    };
+    runNextCommand(commands);
   });
 };
 
 /**
- * End the node process with the given error message and code.
- * Ensures an exit code of at least one, in case the error doesn't have a code.
- * @param {Error} err
+ * Execute a list of shell commands in parallel.
+ * @param {string[]} commands - array of shell commands to be executed in parallel.
+ * @return {Promise.<string[]>} which resolves to an array of stdout results
+ *         from each command when all subprocesses complete, or rejects if
+ *         any of them fail.
  */
-exports.exitWithError = function (err) {
-  console.log("\nError: " + err.message);
-  warnIfWrongNodeVersion();
-  process.exit(err.status || 1);
+exports.executeParallel = function (commands) {
+  var options = getChildProcessOptions();
+  return Promise.all(commands.map(function (command) {
+    return new Promise(function (resolve, reject) {
+      console.log(command);
+      child_process.exec(command, options, function (error, stdout) {
+        if (error) {
+          reject(transformExecError(error));
+        } else {
+          resolve(stdout);
+        }
+      });
+    });
+  }));
```

Comment:
Hmm...I wonder if we should just use something like https://github.com/caolan/async. Thoughts?